### DISCUSSION TITLE
Timeouts

### DISCUSSION THREAD

**Discussion by basz** [uyamazak/hc-pdf-server#450] [Q&A] [answered]
Hello, 

I am seeing some timeout errors for a target that is working normally.

Navigation timeout of 30000 ms exceeded

Things to note;

- Runs in docker-compose up stack.
- I always query the same server, different URL.
- I currently don't know when these timeouts begin to happen. But I do see them consistently once they start.
- These timeout's occur immediately for a single request. So it seems the process is confused.
- It seems they happens once and then gets stuck in it's confused state.
- Once I restart the container it is able to work as normal.

Is there something I could tweak or do to figure out what is going on?

I have tried to increase the timeout. Same thing happens.

Note sure if this is the place to ask... Please tell if it is not appropriate here...

part of log.
```
error TimeoutError: Navigation timeout of 30000 ms exceeded
{"level":30,"time":1643969620911,"pid":30,"hostname":"301e7c15652e","reqId":"req-cy","res":{"statusCode":500},"responseTime":13.491924285888672,"msg":"request completed"}
{"level":30,"time":1643969934828,"pid":30,"hostname":"301e7c15652e","reqId":"req-cz","req":{"method":"GET","url":"/?url=https%3A%2F%2Flab.plhw.nl%2Fp%2Fauthenticate-from-token%2FREDACTEDb%3Fredirect_uri%3D%252Forders%252Fview%252Fe2d6be66-19fa-4652-be45-164311a234f5%253FwithStatus%253Dopened%2526withOrderStatus%253Dplaced&pdf_option=A4","hostname":"pdf:8082","remoteAddress":"192.168.224.26","remotePort":56326},"msg":"incoming request"}
error TimeoutError: Navigation timeout of 30000 ms exceeded
{"level":30,"time":1643969934847,"pid":30,"hostname":"301e7c15652e","reqId":"req-cz","res":{"statusCode":500},"responseTime":13.801191329956055,"msg":"request completed"}
{"level":30,"time":1643970512294,"pid":30,"hostname":"301e7c15652e","reqId":"req-d0","req":{"method":"GET","url":"/?url=https%3A%2F%2Flab.plhw.nl%2Fp%2Fauthenticate-from-token%2FREDACTEDb%3Fredirect_uri%3D%252Forders%252Fview%252F38dbfe98-00b0-4141-92db-0aa9ff8af0a7%253FwithStatus%253Dopened%2526withOrderStatus%253Dplaced&pdf_option=A4","hostname":"pdf:8082","remoteAddress":"192.168.224.26","remotePort":33766},"msg":"incoming request"}
error TimeoutError: Navigation timeout of 30000 ms exceeded
{"level":30,"time":1643970512311,"pid":30,"hostname":"301e7c15652e","reqId":"req-d0","res":{"statusCode":500},"responseTime":12.573944091796875,"msg":"request completed"}
{"level":30,"time":1643971384440,"pid":30,"hostname":"301e7c15652e","reqId":"req-d1","req":{"method":"GET","url":"/?url=https%3A%2F%2Flab.plhw.nl%2Fp%2Fauthenticate-from-token%2FREDACTEDb%3Fredirect_uri%3D%252Forders%252Fview%252F38dbfe98-00b0-4141-92db-0aa9ff8af0a7%253FwithStatus%253Dopened%2526withOrderStatus%253Dplaced&pdf_option=A4","hostname":"pdf:8082","remoteAddress":"192.168.224.26","remotePort":40368},"msg":"incoming request"}
error TimeoutError: Navigation timeout of 30000 ms exceeded
{"level":30,"time":1643971384474,"pid":30,"hostname":"301e7c15652e","reqId":"req-d1","res":{"statusCode":500},"responseTime":29.08257484436035,"msg":"request completed"}
{"level":30,"time":1643971484521,"pid":30,"hostname":"301e7c15652e","reqId":"req-d2","req":{"method":"GET","url":"/?url=https%3A%2F%2Flab.plhw.nl%2Fp%2Fauthenticate-from-token%2FREDACTEDb%3Fredirect_uri%3D%252Forders%252Fview%252F38dbfe98-00b0-4141-92db-0aa9ff8af0a7%253FwithStatus%253Dopened%2526withOrderStatus%253Dplaced&pdf_option=A4","hostname":"pdf:8082","remoteAddress":"192.168.224.26","remotePort":41020},"msg":"incoming request"}
error TimeoutError: Navigation timeout of 30000 ms exceeded
{"level":30,"time":1643971484613,"pid":30,"hostname":"301e7c15652e","reqId":"req-d2","res":{"statusCode":500},"responseTime":91.77993965148926,"msg":"request completed"}

```

**Comment by uyamazak** [comment]
@basz 
Thanks for comment.

I was able to reproduce same problem in localhost by following the steps below.

Add an endpoint to timeout to `src/app.ts`
```
  server.get(`/timeout`, async () => {
    await new Promise((resolve) => setTimeout(resolve, 1000000))
    return 'ok'
  })
```
build and run docker
```
$ docker build -t hc-pdf-server:latest . 
$ docker run -it -p 8080:8080 hc-pdf-server:latest
```

Access and wait for timeout.
http://localhost:8080/?url=http://localhost:8080/timeout

The following log was output in stdout.
```
{"level":30,"time":1644304778166,"pid":31,"hostname":"58b4240e3d1d","msg":"Server listening at http://0.0.0.0:8080"}
Server listening at http://0.0.0.0:8080
{"level":30,"time":1644304778598,"pid":31,"hostname":"58b4240e3d1d","reqId":"req-1","req":{"method":"GET","url":"/?url=http://localhost:8080/timeout","hostname":"localhost:8080","remoteAddress":"172.17.0.1","remotePort":63344},"msg":"incoming request"}
{"level":30,"time":1644304778637,"pid":31,"hostname":"58b4240e3d1d","reqId":"req-2","req":{"method":"GET","url":"/timeout","hostname":"localhost:8080","remoteAddress":"127.0.0.1","remotePort":42344},"msg":"incoming request"}
{"level":30,"time":1644304782491,"pid":31,"hostname":"58b4240e3d1d","reqId":"req-3","req":{"method":"GET","url":"/?url=http://localhost:8080/timeout","hostname":"localhost:8080","remoteAddress":"172.17.0.1","remotePort":63346},"msg":"incoming request"}
{"level":30,"time":1644304802483,"pid":31,"hostname":"58b4240e3d1d","reqId":"req-4","req":{"method":"GET","url":"/timeout","hostname":"localhost:8080","remoteAddress":"127.0.0.1","remotePort":42350},"msg":"incoming request"}
error TimeoutError: Navigation timeout of 30000 ms exceeded
error TimeoutError: Navigation timeout of 30000 ms exceeded
{"level":30,"time":1644304812477,"pid":31,"hostname":"58b4240e3d1d","reqId":"req-3","res":{"statusCode":500},"responseTime":30006.811861000024,"msg":"request completed"}
```

After a few same accesses (May be HCPDF_PAGES_NUM: default 3), got the following error response immediately.

```
{"error":{"name":"TimeoutError"},"url":"http://localhost:8080/timeout"}
```

When this happens, changing url will always return the same error.

http://localhost:8080/?url=https://www.google.com/
```
{"error":{"name":"TimeoutError"},"url":"https://www.google.com/"}
```

It seems that the error status has been left in the Puppeteer's Page.
I will think about the countermeasure.

**Comment by basz** [comment]
To add to this. 

As a work around I added a hard process.exit() to app.ts when a timeout occurs. Docker restarts the process automatically, so I thought that would be a way to reset the puppeteer. However I have found that that when I do a request for the same url the timeout will still occurs. As soon a I change the url a bit the request goes through as expected.

So. for now I will add &_=random_value to each requested url and see if that indeed resolves it. Weird. Seems like a docker thing or perhaps a memory leak.

**Comment by uyamazak** [comment]
I'm not sure if it's exactly the same as your problem, 
but I was able to reproduce the failure caused by the Timeout Error on the test, and I was able to fix it in below PR.

https://github.com/uyamazak/hc-pdf-server/pull/457

that test failed with installing @uyamazak/fastify-hc-pages@1.0.6
[test/requests/get-timeout.test.ts](https://github.com/uyamazak/hc-pdf-server/pull/457/files#diff-3b56fb24ea4f5fd0f39be8944bd36fdb3e8e3811ff99c6e6f24ceaf4db82efc7)

If possible, I would appreciate it if you could try it in your environment.

**Comment by basz** [comment]
HI, circling back. I applied these changes and things looked better. However I still got unexpected and seemingly random timeouts. Upon closer inspection it turned out a to be an issue in some of the target url's.

Some of them were rendering STL files with threejs. And that is using WebGL to do that. It took me some time to realise that WebGL is isn't active in chrome in puppeteer. duh.

thanks for you eyes on this. great product.

**Comment by uyamazak** [comment, marked as answer]
@basz 
Thanks for the confirmation and the detailed cause of the timeout!

>WebGL to do that. It took me some time to realise that WebGL is isn't active in chrome in puppeteer.

WebGL may be the cause of the problem, and maybe the following settings that are hard-coded may be the problem.
I'll add an environment variable to switch it later.

https://github.com/uyamazak/fastify-hc-pages/blob/main/src/hc-pages.ts#L18

**Comment by basz** [comment]
When ever i start chrome as follows `open -a "Google Chrome" --args  -disable-webgl ` (MacOS). And then visit STL pages I get normal exceptions which I can catch.

With this things just seem to hang. There is no response. 

I am curious about that --disable-gpu flag though. would it not hang? I'm sure it can;t rendered via webgl without a gpu.

Anyway ThreeJS has a detection for the presence of webGL but will incorrectly report availability eventually still resulting in a hang.

For now I just changed my app to look at the url and specific GET var is present no stl's are rendered.

**Comment by uyamazak** [comment]
Thanks for the detailed info.
I have just released the new version that adds the ability to change that --args of Browser Launch Options.

https://github.com/uyamazak/hc-pdf-server#hcep_default_browser_launch_args
